### PR TITLE
Adding service for accessing ESA hubble data

### PR DESCRIPTION
**Pull request by esdc-esac-esa-int**
This is a replacement for the old PR #1345 after cleaning up

### CODE REVIEW

**Comment by esdc-esac-esa-int** [comment]
I reproduce the PR cleaning up all changes. I hope it is ok now. Only missing the milestone which was introduced by @keflavich in the old PR and the change log. I updated the CHANGE file but it seems it is not enough. I cannot see the error message in full: "Cannot check for consistency of milestone since m..."

**Comment by bsipocz** [comment]
@esdc-esac-esa-int - there is something weird going on in here. The first commit somehow got added as I being the committer (while wasn't) and it still contains all the unrelated helpers, etc changes. The merge commit later on covers it up so they don't show in the diff on github.
Given that the other more detailed branch is now deleted I'm not able to go back and do a rebase and squash based on that. However it should be possible as the owner of the fork to restore a branch, could you do that please?

File: astroquery/esa_hubble/tests/dummy_handler.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            raise Exception("".join(("Method '",
+                                     str(method),
+                                     "' not invoked. (Invoked method is '",
+                                     str(self._invokedMethod)+"')")))
+
+    def check_parameters(self, parameters, method_name):
+        if parameters is None:
+            return len(self._parameters) == 0
+            if len(parameters) != len(self._parameters):
+                raise Exception("Wrong number of parameters for method '%s'. \
```
Comment (bsipocz) [inline astroquery/esa_hubble/tests/dummy_handler.py]: Any raised exception needs to be more specific than the most generic `Exception`

File: astroquery/esa_hubble/tests/test_esa_hubble.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+
+@author: Javier Duran
+@contact: javier.duran@sciops.esa.int
+
+European Space Astronomy Centre (ESAC)
+European Space Agency (ESA)
+
+Created on 13 Ago. 2018
```
Comment (bsipocz) [inline astroquery/esa_hubble/tests/test_esa_hubble.py]: Keep English the language

File: astroquery/esa_hubble/tests/test_esa_hubble.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+
+"""
+import pytest
+
+from astroquery.esa_hubble.core import ESAHubbleClass
+from astroquery.esa_hubble.tests.dummy_handler import DummyHandler
+from astroquery.esa_hubble.tests.dummy_tap_handler import DummyHubbleTapHandler
+from astropy import coordinates
+# from astropy.tests.helper import remote_data
```
Comment (bsipocz) [inline astroquery/esa_hubble/tests/test_esa_hubble.py]: remove unused code

File: docs/esa_hubble/esa_hubble.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
+.. doctest-skip-all
+
+.. _astroquery.esa_hubble:
+
+************************************
+esa_hubble (`astroquery.esa_hubble`)
+************************************
+
+The Hubble Space Telescope (HST) is a joint ESA/NASA orbiting astronomical observatory operating from the near-infrared into the ultraviolet.  Launched in 1990 and scheduled to operate at least through 2020, HST carries and has carried a wide variety of instruments producing imaging, spectrographic, astrometric, and photometric data through both pointed and parallel observing programs. During its lifetime HST has become one of the most important science projects ever, with over 500 000 observations of more than 30000 targets available for retrieval from the Archive.
```
Comment (bsipocz) [inline docs/esa_hubble/esa_hubble.rst]: wrap long lines

File: docs/esa_hubble/esa_hubble.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+This package allows the access to the European Space Agency Hubble Archive (EHST; http://archives.esac.esa.int/ehst/). All the HST observations available in the EHST are synchronised with the MAST services for HST reprocessed public data and corresponding metadata. Therefore, excluding proprietary data, all HST data in the EHST are identical to those in MAST.
```
Comment (bsipocz) [inline docs/esa_hubble/esa_hubble.rst]: links should be hyperlinks not just normal text

File: astroquery/esa_hubble/tests/test_esa_hubble.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        dummyHandler = DummyHandler("launch_job", parameters)
+        dummyTapHandler = DummyHubbleTapHandler("get_tables", parameters2)
+        ehst = ESAHubbleClass(dummyHandler, self.get_dummy_tap_handler())
+        ehst.get_tables(True, True)
+        dummyTapHandler.check_call("get_tables", parameters2)
+
+    def test_get_columns(self):
+        parameters = {}
+        parameters['query'] = "select top 10 * from hsc_v2.hubble_sc2"
```
Comment (bsipocz) [inline astroquery/esa_hubble/tests/test_esa_hubble.py]: this is a weird way to populate a dict, why not do it in one step?

File: astroquery/esa_hubble/tests/dummy_handler.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import os
+from astroquery.utils.tap.model import modelutils, taptable
+from requests.models import Response
+
+__all__ = ['ESAHubble', 'ESAHubbleClass', 'Conf', 'conf',
```
Comment (bsipocz) [inline astroquery/esa_hubble/tests/dummy_handler.py]: This is not right, most of these are not defined in this file

File: astroquery/esa_hubble/core.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+class ESAHubbleHandler(BaseQuery):
+
+    def __init__(self):
+        super(ESAHubbleHandler, self).__init__()
+
+    def get_file(self, filename, response, verbose=False):
+        with open(filename, 'wb') as fh:
+            fh.write(response.content)
+
+        if "/" not in filename:
```
Comment (bsipocz) [inline astroquery/esa_hubble/core.py]: would this work on windows?

File: astroquery/esa_hubble/core.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+        """
+
+        obs_id = "OBSERVATION_ID=" + observation_id
+        cal_level = "CALIBRATION_LEVEL=" + calibration_level
+        link = self.data_url + obs_id + "&" + cal_level
+
+        if filename is None:
+            filename = observation_id + ".tar"
+
+        # response = self._download_file(link, local_filepath=filename,
```
Comment (bsipocz) [inline astroquery/esa_hubble/core.py]: remove unused code

File: astroquery/esa_hubble/core.py [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+                   ".RA,POSITION.DEC,POSITION."
+                   "GAL_LAT,POSITION.GAL_LON,POSITION.ECL_LAT,POSITION.ECL_LON"
+                   ",POSITION.FOV_SIZE,ENERGY."
+                   "WAVE_CENTRAL,ENERGY.WAVE_BANDWIDTH,ENERGY.WAVE_MAX,ENERGY"
+                   ".WAVE_MIN,ENERGY.FILTER FROM"
+                   " FIELD_NOT_USED  WHERE OBSERVATION.COLLECTION='HST'  AND  "
+                   "PLANE.MAIN_SCIENCE_PLANE="
+                   "'true'  AND  (OBSERVATION.TYPE='HST Composite' OR "
+                   "OBSERVATION.TYPE='HST Singleton')"
+                   "  AND  INTERSECTS(CIRCLE('ICRS'," + str(ra) + ""
```
Comment (bsipocz) [inline astroquery/esa_hubble/core.py]: why not use string formatting?

File: astroquery/esa_hubble/core.py [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+            return table_names
+        else:
+            return tables
+
+    def get_columns(self, table_name=None, only_names=True, verbose=False):
+        """Get the available columns for a table in EHST TAP service
+
+        Parameters
+        ----------
+        table_name : string, mandatory, default None
```
Comment (bsipocz) [inline astroquery/esa_hubble/core.py]: Why make it optional if it's in fact required?

File: astroquery/esa_hubble/__init__.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+=============
+ehst
```
Comment (bsipocz) [inline astroquery/esa_hubble/__init__.py]: What does this mean?

**Comment by bsipocz** [comment]
Now I've rebased this and pushed back to the branch, no need to revert the other one any more. 

Here is a super good resource to exercise a bit the branching workflow as with this module it already derailed a few times: https://learngitbranching.js.org/

File: astroquery/esa_hubble/core.py (hunk 8 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa_hubble/core.py]: This change was introduced by @keflavich if I recall correctly. Should I go back to the previous way?

Comment (keflavich) [reply astroquery/esa_hubble/core.py]: I don't recall doing this, but perhaps I did.  In any case, it's not very important functionality, as it's just governing what log message is reported.

Comment (esdc-esac-esa-int) [reply astroquery/esa_hubble/core.py]: Maybe @bsipocz is referring here only to the "/" part. I can modify it if you prefer by os.pathsep

**Comment by esdc-esac-esa-int** [comment]
How can I do to make it merged?

**Comment by esdc-esac-esa-int** [comment]
I think I addressed all comment. Let me know if that is not the case.

BTW, do you know by when the new version will be released?

**Comment by bsipocz** [comment]
@esdc-esac-esa-int - this PR is still full of merge commits that contain unrelated chaneges. I've already rebased it a few times, but the merges get pushed back. 

Also, I would think that a new release won't be tagged before doing some restructuring of the namespace as it was discussed before.

**Comment by keflavich** [comment]
@esdc-esac-esa-int The problem is the merge commits that look like this:
image
This causes problems with our commit history that we do not want to add to the repository.  We need the commit history of this pull request (_not_ a new pull request, please!) to not have any merge commits.

**Comment by esdc-esac-esa-int** [comment]
I identified the following merge commits in my log (git log):

```
merge branch 'master' of https://github.com/astropy/astroquery into esa_hubble

commit becfd5ea4ab2f641f2d25323235ad54c704ad3c7
Merge: 196f4b95 61561509
Author: ESAC Science Data Centre <archives.esac.esa.int@gmail.com>
Date:   Mon Mar 4 11:36:29 2019 +0100

    Merge branch 'master' into esa_hubble

…

commit 6f85c9dc5f6eb75635100b918918083da9056369
Author: jduran <jduran@sciops.esa.int>
Date:   Wed Feb 27 17:43:27 2019 -0800

    Adding initial commit for esa_hubble, rebased by bsipocz but manually set author back to jduran

commit 554f2aa348107f971f6f20ce7a8c1da6a01b63c5
Merge: ad8601e3 91a9ef56
Author: jduran <jduran@sciops.esa.int>
Date:   Wed Feb 27 14:10:17 2019 +0100

    Merge branch 'master' of https://github.com/astropy/astroquery into esa_hubble
```
In order to solve this, I plan to do the following:
```
git reset --hard <commit_id>
git reset --soft HEAD@{1}
git commit -m "Reverting to the state of the project at <commit_id>"
```
Is this correct? I am assuming that I have to do it with the first of the merge commits (554f2aa348107f971f6f20ce7a8c1da6a01b63c5)

**Comment by keflavich** [comment]
I actually don't know how to handle this situation.  There was originally a long list of commits that got squashed into just a few commits.

**Comment by esdc-esac-esa-int** [comment]
What can we do now then?

**Comment by bsipocz** [comment]
@esdc-esac-esa-int - What was needed is a `git fetch --all`, then a `git rebase -i astroquery/master`, then delete all but the 4 commits that has the actual changes. As it was said before I strongly recommend to check out the git tutotials linked from the astropy development workflow, especially the one on working with branches (https://learngitbranching.js.org/). http://docs.astropy.org/en/latest/development/workflow/development_workflow.html#new-to-git

**Comment by bsipocz** [comment]
@esdc-esac-esa-int - there are no tests that actually query the remote service. They need to be added.

**Comment by esdc-esac-esa-int** [comment]
I performed the actions you asked me to do @bsipocz on the branch. Is it ok now?

**Comment by bsipocz** [comment]
yes, thanks for adding the remote data tests. However this needs the rebase again, there is merge commit in there again, as well as `astropy_helpers` got deleted. 
If you happen to be in the IVOA meeting I'm happy to walk you through how a rebase should be done during the hackdays.

**Comment by esdc-esac-esa-int** [comment]
Hi @bsipocz, I am afraid I won't be attending this hackathon. @jcsegovia will be. Juan Carlos Segovia is a colleague of mine, you can talk to him.

**Comment by bsipocz** [comment]
@esdc-esac-esa-int - I'll rebase and do the minimal cleanup before merge. With @jcsegovia we went through a rather painful rebase session at the meeting, I'm sure he has enough insight to the process to show you the tricks next time.

**Comment by bsipocz** [comment]
Thanks @jduran!

**Pull request merged**